chain-signatures/node/src/mesh/connection.rs

```diff
@@ -60,41 +62,77 @@ impl Pool {
         for (participant, info) in connections.iter() {
-            match self.fetch_participant_state(info).await {
-                Ok(state) => match self.send_empty_msg(participant, info).await {
-                    Ok(()) => {
-                        status.insert(*participant, state);
-                        participants.insert(participant, info.clone());
-                    }
+            let participant = *participant;
+            let info = info.clone();
+            let self_clone = Arc::clone(&self); // Clone Arc for use inside tasks
+
+            join_set.spawn(async move {
+                match self_clone.fetch_participant_state(&info).await {
+                    Ok(state) => match self_clone.send_empty_msg(&participant, &info).await {
+                        Ok(()) => Ok((participant, state, info)),
+                        Err(e) => {
+                            tracing::warn!(
+                                "Send empty msg for participant {participant:?} with url {} has failed with error {e}.",
+                                info.url
+                            );
+                            Err(())
+                        }
+                    },
                     Err(e) => {
-                        tracing::warn!("Send empty msg for participant {participant:?} with url {} has failed with error {e}.", info.url);
+                        tracing::warn!(
+                            "Fetch state for participant {participant:?} with url {} has failed with error {e}.",
+                            info.url
+                        );
+                        Err(())
                     }
-                },
+                }
+            });
+        }
+
+        let mut status = self.status.write().await;
+        let mut participants = Participants::default();
+
+        // Process completed tasks
+        while let Some(result) = join_set.join_next().await {
+            match result {
+                Ok(Ok((participant, state, info))) => {
+                    status.insert(participant, state);
+                    participants.insert(&participant, info);
+                }
+                Ok(Err(())) => {
+                    // Already logged in task
+                }
                 Err(e) => {
-                    tracing::warn!("Fetch state for participant {participant:?} with url {} has failed with error {e}.", info.url);
+                    tracing::warn!("fetch participant state task panicked: {e}");
                 }
             }
         }
+
         drop(status);
 
+        // Update the active participants
         let mut active = self.current_active.write().await;
         *active = Some((participants.clone(), Instant::now()));
+
         participants
     }
 
-    pub async fn ping_potential(&self) -> Participants {
+    // self typed Arc<Self> so it can be passed between tokio tasks
+    pub async fn ping_potential(self: Arc<Self>) -> Participants {
```

Comment:
looks like some logics can be reused with `ping`, but we can do in a future refactor PR